### PR TITLE
KCRM updates and improvements

### PR DESCRIPTION
**Pull request by jhennawi**
This PR reflects efforts by @jhennawi and @rcooke-ast to improve support for KCRM. This is a placeholder for the moment, but the version log and description of the PR will be populated shortly.

### CODE REVIEW

**Comment by rcooke-ast** [comment]
As a heads-up to @profxj and @kbwestfall - we should all hold off reviewing until PR #1725 is merged.

File: pypeit/find_objects.py
Comment (profxj) [inline pypeit/find_objects.py]: remove?

File: pypeit/find_objects.py
Comment (profxj) [inline pypeit/find_objects.py]: now or never?

File: pypeit/find_objects.py
Comment (profxj) [inline pypeit/find_objects.py]: move to `deprecate/` ?

File: pypeit/images/rawimage.py
Comment (profxj) [inline pypeit/images/rawimage.py]: remove?

File: pypeit/par/pypeitpar.py
Comment (profxj) [inline pypeit/par/pypeitpar.py]: do now?

File: pypeit/spectrographs/keck_deimos.py
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py:768]: maybe remove this and leave for the forthcoming DEIMOS PR?

**Review by profxj — APPROVED** [review]
nice work folks!

File: pypeit/core/datacube.py
Comment (rcooke-ast) [inline pypeit/core/datacube.py:1235]: Agreed - let's remove the commented text

File: pypeit/core/datacube.py
Comment (rcooke-ast) [inline pypeit/core/datacube.py]: Added...

File: pypeit/core/datacube.py
Comment (rcooke-ast) [inline pypeit/core/datacube.py]: Done.

File: pypeit/extraction.py
Comment (rcooke-ast) [inline pypeit/extraction.py:891]: I'm still against the idea that we should extract by default. It would be good to have some flexibility here. Perhaps we can allow extractions for red optical and NIR IFUs, but not for KCWI (blue)?

Should we add in a par that specifies if the 1D extraction is to be performed? And add docs in KCWI (blue) to caution folks not to use the extracted objects for their science?

File: pypeit/images/rawimage.py
Comment (rcooke-ast) [reply pypeit/images/rawimage.py]: Removed - it's inside a debug, but I don't think it's needed...

File: pypeit/par/pypeitpar.py
Comment (rcooke-ast) [reply pypeit/par/pypeitpar.py]: Done

File: pypeit/find_objects.py
Comment (rcooke-ast) [reply pypeit/find_objects.py]: Agreed - removed.

File: pypeit/find_objects.py
Comment (rcooke-ast) [inline pypeit/find_objects.py]: It is necessary for the joint skysub (you need the wavelength to be correct to fit all slices simultaneously). The idea is that a model of the sky spectrum is computed, and this is used to perform a relative flexure of all slices. Then, the "absolute" flexure is computed with a single (reference) slice and applied to all slices.

I think it can be (eventually) moved to the flexure code, but that needs a lot of tidying up first.

File: pypeit/find_objects.py
Comment (rcooke-ast) [reply pypeit/find_objects.py]: Moved

File: pypeit/find_objects.py
Comment (rcooke-ast) [reply pypeit/find_objects.py]: Moved to deprecated

File: pypeit/find_objects.py
Comment (rcooke-ast) [inline pypeit/find_objects.py]: I've made some changes to the routine, so nothing gets assigned in here. I've also updated the docs so that it's clear what is being calculated and returned.

I think the flexure correction code is in desperate need of being updated. It needs a DataModel and a lot of reworking. It's a big job. For now I propose to keep this here, but it should be included as part of a future improvement to the flexure code.

**Review by rcooke-ast — COMMENTED** [review]
Several comments and responses added

File: pypeit/alignframe.py
Comment (kbwestfall) [inline pypeit/alignframe.py]: This construction feels very cumbersome.  I've implemented a change in the `kcrm_dec_2023_kbw` branch.  Please have a look:

https://github.com/pypeit/PypeIt/compare/kcrm_dec_2023...kcrm_dec_2023_kbw

File: pypeit/coadd3d.py
Comment (kbwestfall) [inline pypeit/coadd3d.py]: Flagging this TODO.

File: pypeit/coadd3d.py
Comment (kbwestfall) [inline pypeit/coadd3d.py]: Combine these 4 double if statements into 4 single if statements?  E.g.:

`if dec_offsets is not None and len(dec_offsets) != self.numfiles:`

File: pypeit/coadd3d.py
Comment (kbwestfall) [inline pypeit/coadd3d.py:446]: In these two instances when you're effectively changing a `cubepar` parameter, can this instead be done when the `ParSet` is validated?

File: pypeit/coadd3d.py
Comment (kbwestfall) [inline pypeit/coadd3d.py]: I'm not sure if either of these is more efficient, but what about:
`self.ra_offsets = [0.0]*self.numfiles`
or
`self.ra_offsets = np.zeros(self.numfiles, dtype=float).tolist()`

File: pypeit/coadd3d.py
Comment (kbwestfall) [inline pypeit/coadd3d.py]: Similar comment about using a for loop (list comprehension) to instantiate the list.

File: pypeit/flatfield.py
Comment (kbwestfall) [inline pypeit/flatfield.py]: remove commented code?

File: pypeit/flatfield.py
Comment (kbwestfall) [inline pypeit/flatfield.py]: Move this TODO to a comment outside the docstring.

File: pypeit/flatfield.py
Comment (kbwestfall) [inline pypeit/flatfield.py]: Is it possible for this to be user error, or is this an edge case that you as the developer expect should never happen?  If the latter, I would put `CODING ERROR` at the beginning of the error statement.

File: pypeit/flatfield.py
Comment (kbwestfall) [inline pypeit/flatfield.py]: Same as code just below.  Turn this into a short utility function?

File: pypeit/flatfield.py
Comment (kbwestfall) [inline pypeit/flatfield.py]: Use `|=` ?

File: pypeit/images/rawimage.py
Comment (kbwestfall) [inline pypeit/images/rawimage.py]: I think this debug block should be abstracted or removed.

File: pypeit/inputfiles.py
Comment (kbwestfall) [inline pypeit/inputfiles.py]: Is this ever true?  Or should you first check if `off_ra` is None before doing the list conversion?

File: pypeit/inputfiles.py
Comment (kbwestfall) [inline pypeit/inputfiles.py]: Same question(s) here.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py:768]: Agreed.

**Review by kbwestfall — CHANGES_REQUESTED** [review]
Looks good!  I have mostly minor requests.

File: pypeit/alignframe.py
Comment (rcooke-ast) [reply pypeit/alignframe.py]: This looks like a good addition to me, and thanks for the added test. I'll leave it to @jhennawi to confirm if this captures what he wanted to achieve.

File: pypeit/coadd3d.py
Comment (rcooke-ast) [reply pypeit/coadd3d.py]: Good point - done!

File: pypeit/coadd3d.py
Comment (rcooke-ast) [reply pypeit/coadd3d.py]: Yeah - the first one is much cleaner

File: pypeit/coadd3d.py
Comment (rcooke-ast) [reply pypeit/coadd3d.py]: Done!

File: pypeit/flatfield.py
Comment (rcooke-ast) [reply pypeit/flatfield.py]: Yes - removed

File: pypeit/flatfield.py
Comment (rcooke-ast) [reply pypeit/flatfield.py]: Done.

File: pypeit/flatfield.py
Comment (rcooke-ast) [reply pypeit/flatfield.py]: Good call - done! I've added this to the core flat code.

File: pypeit/flatfield.py
Comment (rcooke-ast) [reply pypeit/flatfield.py]: Sounds good - added

File: pypeit/flatfield.py
Comment (rcooke-ast) [reply pypeit/flatfield.py]: Agreed - much cleaner

File: pypeit/images/rawimage.py
Comment (rcooke-ast) [reply pypeit/images/rawimage.py]: Yeah, you're right... this was just me testing that the flatfield is working correctly. It's been useful for debugging this PR, and I intended to remove it - now removed.

File: pypeit/inputfiles.py
Comment (rcooke-ast) [reply pypeit/inputfiles.py]: No - I've removed this... it's leftover from the path/files search "bug" that was fixed in this PR.

File: pypeit/inputfiles.py
Comment (rcooke-ast) [reply pypeit/inputfiles.py]: Same as above - fixed!

File: pypeit/spectrographs/keck_deimos.py
Comment (rcooke-ast) [reply pypeit/spectrographs/keck_deimos.py:768]: I leave this for @jhennawi

File: pypeit/coadd3d.py
Comment (rcooke-ast) [reply pypeit/coadd3d.py]: I'll address this once we decide if this will be in 1.15 or not

File: pypeit/coadd3d.py
Comment (rcooke-ast) [reply pypeit/coadd3d.py:446]: I don't think so (because `ra_offsets` is (usually) expected to be a parameter in the `spec2d` data block).

**Review by rcooke-ast — COMMENTED** [review]
Addressed the comments from @kbwestfall - thanks!

File: pypeit/alignframe.py
Comment (kbwestfall) [reply pypeit/alignframe.py]: Okay, issued #1757 and leave it to @jhennawi to approve.

**Comment by rcooke-ast** [comment]
Tests are currently looking good. The one vet test failure is Keck HIRES.
image

File: pypeit/coadd3d.py
Comment (rcooke-ast) [reply pypeit/coadd3d.py]: Now done

**Review by kbwestfall — APPROVED** [review]

**Comment by kbwestfall** [comment]
Tests pass
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  247 passed, 56 warnings in 377.22s (0:06:17) ---
--- PYTEST UNIT TESTS PASSED  133 passed, 171 warnings in 957.98s (0:15:57) ---
--- PYTEST VET TESTS PASSED  61 passed, 70 warnings in 4605.44s (1:16:45) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 237/237 TESTS  ---
Testing Started at 2024-02-03T01:25:57.592117
Testing Completed at 2024-02-03T13:04:37.174274
Total Time: 11:38:39.582157
```
Merging

**Pull request merged**
